ci/cloudbuild/cache.sh

```diff
@@ -0,0 +1,144 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# This script saves a cache tarball to GCS, and restores one from GCS. It is
+# intended to be called from a build step in the `cloudbuild.yaml` file. It
+# executes in the context of build. Typically the "restore cache" step will be
+# called in one step, followed by the actual build of the code, followed by the
+# "save cache" step.
+#
+# Usage: ./cache <save|restore>
+#
+#   Options:
+#     --bucket_url=<url>    URL to the GCS bucket where the cache should be saved
+#     --key=<key>           The primary key to use when caching
+#     --fallback_key=<key>  If the primary key isn't found when restoring, try this key
+#     --path=<path>         A path to be backed up. May be specified multiple times.
+#     -h|--help             Print this help message
+
+set -euo pipefail
+
+source "$(dirname "$0")/../lib/init.sh"
+source module ci/lib/io.sh
+
+function print_usage() {
+  # Extracts the usage from the file comment starting at line 17.
+  sed -n '17,/^$/s/^# \?//p' "${PROGRAM_PATH}"
+}
+
+# Use getopt to parse and normalize all the args.
+PARSED="$(getopt -a \
+  --options="h" \
+  --longoptions="bucket_url:,key:,fallback_key:,path:,help" \
+  --name="${PROGRAM_NAME}" \
+  -- "$@")"
+eval set -- "${PARSED}"
+
+BUCKET_URL=""
+KEY=""
+FALLBACK_KEY=""
+PATHS=()
+while true; do
+  case "$1" in
+  --bucket_url)
+    BUCKET_URL="$2"
+    shift 2
+    ;;
+  --key)
+    KEY="$2"
+    shift 2
+    ;;
+  --fallback_key)
+    FALLBACK_KEY="$2"
+    shift 2
+    ;;
+  --path)
+    PATHS+=("$2")
+    shift 2
+    ;;
+  -h | --help)
+    print_usage
+    exit 0
+    ;;
+  --)
+    shift
+    break
+    ;;
+  esac
+done
+readonly BUCKET_URL
+readonly KEY
+readonly FALLBACK_KEY
+readonly PATHS
+
+if [[ $# -eq 0 ]]; then
+  echo "Missing action: save|restore"
+  print_usage
+  exit 1
+elif [[ -z "${BUCKET_URL}" ]]; then
+  echo "Missing --bucket_url="
+  print_usage
+  exit 1
+elif [[ -z "${KEY}" ]]; then
+  echo "Missing --key="
+  print_usage
+  exit 1
+fi
+
+readonly PRIMARY_CACHE_URL="${BUCKET_URL}/${KEY}/cache.tar.gz"
+readonly FALLBACK_CACHE_URL="${BUCKET_URL}/${FALLBACK_KEY}/cache.tar.gz"
+
+function save_cache() {
+  paths=()
+  for p in "${PATHS[@]}"; do
+    if [[ -r "${p}" ]]; then paths+=("${p}"); fi
+  done
+  if ((${#paths[@]} == 0)); then
+    io::log "No paths to cache found."
+    return 0
+  fi
+  io::log "Saving ${PATHS[*]} to ${PRIMARY_CACHE_URL}"
+  tar -czf cache.tar.gz "${PATHS[@]}"
+  io::log "  Cache size: $(du -sh cache.tar.gz)"
+  gsutil -m cp cache.tar.gz "${url}"
+}
+
+function restore_cache() {
+  for url in "${PRIMARY_CACHE_URL}" "${FALLBACK_CACHE_URL}"; do
+    if gsutil stat "${url}"; then
+      io::log "Fetching cache url ${url}"
+      gsutil -m cp "${url}" cache.tar.gz || continue
+      io::log "  Cache size: $(du -sh cache.tar.gz)"
+      tar -zxf cache.tar.gz && break
```

Comment:
FWIW, `gsutil cp` supports `-` as a source or destination, so you could do `gsutil cp "${url}" - | tar -zxf -`

Reply:
Thanks. Done.